proto/v1/sql_service.proto

```diff
@@ -14,6 +16,86 @@ service SQLService {
       body: "*"
     };
   }
+  rpc Query(QueryRequest) returns (QueryResponse) {
+    option (google.api.http) = {
+      post: "/v1/{instance=instances/*}:query"
+      body: "*"
+    };
+    option (google.api.method_signature) = "instance";
+  }
+}
+
+message QueryRequest {
+  // The instance to execute the query against.
+  // Format: instances/{instance}
+  string instance = 1 [(google.api.field_behavior) = REQUIRED];
+
+  // The connection database name to execute the query against.
+  // For PostgreSQL, it's required.
+  // For other database engines, it's optional. Use empty string to execute against without specifying a database.
+  string connection_database = 2;
+
+  // The SQL statement to execute.
+  string statement = 3;
+
+  // The maximum number of rows to return.
+  int32 limit = 4;
+}
+
+message QueryResponse {
+  // The query results.
+  repeated QueryResult results = 1;
+
+  // The query advices.
+  repeated Advice advices = 2;
+}
+
+message QueryResult {
+  // Column names of the query result.
+  repeated string column_names = 1;
+
+  // Column types of the query result.
+  repeated string column_type_names = 2;
+
+  // Rows of the query result.
+  repeated QueryRow rows = 3;
+
+  // Columns are masked or not.
+  repeated bool masked = 4;
+
+  // The error message if the query failed.
+  string error = 5;
+}
+
+message QueryRow {
+  // Row values of the query result.
+  repeated string values = 1;
+}
+
+enum Status {
+  STATUS_SUCCESS = 0;
+  STATUS_WARN = 1;
+  STATUS_ERROR = 2;
+}
+
+message Advice {
+  // The advice status.
+  Status status = 1;
```

Comment:
We can embed `Status` into `Advice`, because there are many things called "Status" out there.